src/main/java/core/Game.java

```diff
@@ -495,13 +502,93 @@ public final void run() {
+            gameState.getGameParameters().randomize();
+            System.out.println("Game parameters: " + gameState.getGameParameters());
+        }
+
+        int lastPlayer = -1; // initialise with no last player, since we are starting a new game
+        int nActionsPerTurn = 1; // keep track within this scope to avoid parallel processes to modify this game's stats
+
+        boolean firstEnd = true;
+//        System.out.println("Running game: "+matchUpPlayers);
+        while (gameState.isNotTerminal()) {
+            int activePlayer = gameState.getCurrentPlayer();
+
+            AbstractPlayer currentPlayer = players.get(activePlayer);
+            /*
+             * The Game is responsible for tracking the players and the current game state
+             * It is important that the Game never passes the main AbstractGameState to the individual players,
+             * but instead always uses copy(playerId) to both:
+             * i) shuffle any hidden data they cannot see
+             * ii) ensure that any changes the player makes to the game state do not affect the genuine game state
+             *
+             * Players should never have access to the Game, or the main AbstractGameState, or to each other!
+             */
+
+            // Get player to ask for actions next (This horrendous line is for backwards compatibility).
+            boolean reacting = gameState instanceof AbstractGameStateWithTurnOrder &&
+                               ((AbstractGameStateWithTurnOrder) gameState).getTurnOrder() instanceof ReactiveTurnOrder &&
+                               !((ReactiveTurnOrder) ((AbstractGameStateWithTurnOrder) gameState).getTurnOrder()).getReactivePlayers().isEmpty();
+
+            // Check if this is the same player as last, count number of actions per turn
+            if (!reacting) {
+                if (currentPlayer != null && activePlayer == lastPlayer) {
+                    nActionsPerTurn++;
+                } else {
+                    nActionsPerTurnSum += nActionsPerTurn; // atomic
+                    nActionsPerTurn = 1;
+                    nActionsPerTurnCount++; // atomic
+                }
+            }
+
+            if (gameState.isNotTerminal()) {
+                if (debug) System.out.printf("Invoking oneAction from Game for player %d%n", activePlayer);
+                // keep track of last player within this scope, since parallel processes may modify this.lastPlayer
+                lastPlayer = gameState.getCurrentPlayer();
+                oneAction(gameState, forwardModel, players);
+            } else {
+                if (firstEnd) {
+                    if (gameState.coreGameParameters.verbose) {
+                        System.out.println("Ended");
+                    }
+                    terminate(gameState, forwardModel);
+                    firstEnd = false;
+                }
+            }
+            if (debug) System.out.println("Exiting synchronized block in Game");
```

Comment:
I kept this debug message in this version of the method, for consistency with debug output. If desired I could change this to something else, or remove it.